### ISSUE TITLE
wrong_self_convention: FN with `is_*` taking `&mut self`

### ISSUE THREAD

**Issue by jendrikw** [rust-lang/rust-clippy#8142]
### Summary

My interpretation of the table at https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention is that a method called `is_*` is allowed to take no self, or `&self`, but not `self` or `&mut self`. In reality, only `self` produces a warning. Is this intentional?

### Lint Name

wrong_self_convention

### Reproducer

I tried this code:

```rust
#![allow(unused)]
struct S { field: String }

impl S {
    fn is_foo(&mut self) -> bool { todo!() }
}

fn main() {}
```

I expected to see this happen:
```
warning: methods called `is_*` usually take `self` by reference or no `self`
  --> src/main.rs:16:15
   |
16 |     fn is_foo(&mut self) -> bool {
   |               ^^^^
```

Instead, this happened:

No warning

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```

**Comment by jendrikw** [comment]
This issue was files because intellij-rust warns for both `&mut self` and `self` and it wants to be consistent with clippy. See intellij-rust/intellij-rust#7557.

![image](https://user-images.githubusercontent.com/12938023/146641625-b6b9706c-0956-4e1c-8953-156ee7a53d0a.png)
(The yellow background is a warning from intellij-rust, the yellow underline is from clippy.)

**Comment by mikerite** [comment]
I agree. It should warn on `&mut self`.

**Comment by nmathewson** [comment]
This one looks neat, and I _think_ I can see what the trouble is; I'll try writing a patch for this in the next day or two.  (If I haven't written a PR by 5 January, I have probably given up or forgotten this.)
